conda-store-server/conda_store_server/_internal/settings.py

```diff
@@ -0,0 +1,170 @@
+# Copyright (c) conda-store development team. All rights reserved.
+# Use of this source code is governed by a BSD-style
+# license that can be found in the LICENSE file.
+
+import functools
+from typing import Any, Callable, Dict
+
+import pydantic
+from sqlalchemy.orm import Session
+
+from conda_store_server import api
+from conda_store_server._internal import schema
+
+
+class Settings:
+    def __init__(self, db: Session, deployment_default: schema.Settings):
+        self.db = db
+        self.deployment_default = deployment_default.model_dump()
+
+    def _ensure_closed_session(func: Callable):
+        @functools.wraps(func)
+        def wrapper(self, *args, **kwargs):
+            result = func(self, *args, **kwargs)
+            self.db.close()
+            return result
+
+        return wrapper
+
+    @_ensure_closed_session
+    def set_settings(
+        self,
+        namespace: str = None,
+        environment_name: str = None,
+        data: Dict[str, Any] = {},
+    ):
+        """Persist settings to the database
+
+        Parameters
+        ----------
+        namespace : str, optional
+            namespace to use to retrieve settings
+        environment_name : str, optional
+            environment to use to retrieve settings
+        data : dict
+            settings to upsert
+        """
+        setting_keys = schema.Settings.model_fields.keys()
+        if not data.keys() <= setting_keys:
+            invalid_keys = data.keys() - setting_keys
+            raise ValueError(f"Invalid setting keys {invalid_keys}")
+
+        for key, value in data.items():
+            field = schema.Settings.model_fields[key]
+            global_setting = field.json_schema_extra["metadata"]["global"]
+            if global_setting and (
+                namespace is not None or environment_name is not None
+            ):
+                raise ValueError(
+                    f"Setting {key} is a global setting cannot be set within namespace or environment"
+                )
+
+            try:
+                validator = pydantic.TypeAdapter(field.annotation)
+                validator.validate_python(value)
+            except Exception as e:
+                raise ValueError(
+                    f"Invalid parsing of setting {key} expected type {field.annotation} ran into error {e}"
+                )
+
+        if namespace is not None and environment_name is not None:
+            prefix = f"setting/{namespace}/{environment_name}"
+        elif namespace is not None:
+            prefix = f"setting/{namespace}"
+        else:
+            prefix = "setting"
```

Comment:
I think the setting can inadvertently be set at the global level if `namespace is None` and `environment_name is not None`. I think something similar to the prefix logic in `get_settings` would work though.

Reply:
oh, good catch. I think that is probably a programmer error and this function should probably throw an exception. Will also add a test for that!